### PR TITLE
Feat/finish new designs migration

### PR DESCRIPTION
**Pull request by dragonflyfree**
## Checklist
- [x] I'm using the correct PHP Version (8.1 for current, 7.4 for legacy).
- [x] I've added comments to any new methods I've created or where else relevant.
- [x] I've replaced magic method usage on DbService classes with the getInstance() static method.
- [x] I've written any documentation for new features or where else relevant in the docs [repo](https://github.com/2pisoftware/cmfive-docs).

## Description

Various changes to Tag and Form modules, changes to Bootstrap5 code, Vue2 to Vue3 (yet to be complete for Form module, may be required in Admin module, for example in the users list? perhaps?) conversion, etc.

## Changelog

refs:
issues:

## Other Information

Docs pull request:

### CODE REVIEW

**Comment by MaddyUnderStars** [comment]
Below is a copy of my last PR #384 as this one is replacing it.
It's strictly *my* work on this migration as I'm not completely aware of what Adam and Claire have done.

general goals:
- migrate pages to bootstrap5 theme.
- - this theme does not include jquery, so it also requires rewriting a lot of systems to use plain js
- while I was writing something due to prev point, I might as well fix some bugs or improve small things
- lastly, if a page was written in plain html, it was easier to just make use of the form generation code or such in php since that already has all the bootstrap5 classes.

modules migrated:
- tag
- task
- timelog
- favourites

by page:
- /task/index ( dashboard )
- /task-group/viewtaskgrouptypes ( manage task groups )
- - uses quill editor instance of a plain text area
- /task/taskweek ( activity )
- /task/tasklist ( task list )
- /task/edit ( task view/edit )
- - only the edit/view and timelog tabs have been migrated currently, as the others are partials
- /task-group/viewmembergroup ( task group members/details page )
- /task-group/viewtaskgroup modal ('edit task group' button in viewmembergroup )
- /task/partials/listtasks
- /task/subscriber
- /task/subscriber/add
- /timelog/partial/listtimelog
- /timelog/index
- /timelog/edit
- /timelog/move
- /task-group/movetaskgroup modal
- /task-group/addgroupmembers

jquery:
- new php class Html5Autocomplete for multiselect free-form text input with autocomplete drop downs included with tom select
- removed jquery usage for CKEDITOR from html::form
- rewrite tags module client js to remove jquery dep.
- - tags module is now included in /system/templates/base/ so as to only include a single tags manager per page vs per inclusion of the partial
- - tags module now fetches tags via ajax
- - can now remove or add multiple tags at once via the modal (#250)
- Favourites
- - FavouriteService::getFavouriteButton now deprecated in favour of getBootstrapButton
- - Migrated favourites list modal to bootstrap
- 
# Screenshots and video ( as of 14/10 )

<details>
	<summary>/task/index</summary>
	
</details>

<details>
	<summary>task-group/viewtaskgrouptypes</summary>
	
	
</details>

<details>
	<summary>/task/taskweek</summary>
	
</details>

<details>
	<summary>/task/tasklist</summary>
	
</details>

<details>
	<summary>/task/edit</summary>
	
	
	
	<video src="https://github.com/user-attachments/assets/f58a92f7-646c-4db4-ab7a-dfd1d854abf8"/>
</details>

<details>
	<summary>/task-group/viewmembergroup and /task/listtasks partial</summary>
	
	
</details>

<details>
	<summary>/task-group/viewtaskgroup modal</summary>
	
</details>

<details>
	<summary>/task/subscriber</summary>
	
	
</details>

<details>
	<summary>/task/listtimelog</summary>
	
	
</details>

<details>
	<summary>/timelog/index</summary>
	
</details>

<details>
	<summary>/timelog/edit</summary>
	note: the assigned user field is only visible as admin user, like it is currently
	
</details>

<details>
	<summary>/timelog/move</summary>
	
</details>

<details>
	<summary>/task-group/movetaskgroup modal</summary>
	
	
</details>

<details>
	<summary>/task-group/addgroupmembers modal</summary>
	
</details>

<details>
	<summary>add tags modal</summary>
	<video src="https://github.com/user-attachments/assets/30741181-8c68-405a-9ec7-d7b89dbc6346"/>
</details>

<details>
	<summary>favourites modal</summary>
	
</details>

File: system/modules/task/templates/edit.tpl.php [hunk 1]
```diff
@@ -1,27 +1,32 @@
...
+                    <span class="badge rounded-pill bg-secondary text-light ms-1"><?php echo $timelog_count; ?></span>
+                </a>
             <?php endif; ?>
-            <a href="#internal_comments">Internal Comments <span class='label secondary round cmfive__tab-label cmfive__count-internal_comment_section'></span></a>
-            <a href="#external_comments">External Comments <span class='label secondary round cmfive__tab-label cmfive__count-external_comment_section'></span></a>
-            <a href="#attachments">Attachments <span class='label secondary round cmfive__tab-label cmfive__count-attachment'></span></a>
+
+            <a href="#internal_comments">Internal Comments</a>
+            <a href="#external_comments">External Comments</a>
+            <a href="#attachments">Attachments</a>
```
Comment (MaddyUnderStars) [inline system/modules/task/templates/edit.tpl.php:29]: TODO: add counts to the tab headers like timelog has

File: system/modules/form/templates/application/edit_member.tpl.php [hunk 2]
```diff
@@ -0,0 +1 @@
+<?php echo $form; ?>
```
Comment (adam-buckley) [inline system/modules/form/templates/application/edit_member.tpl.php:1]: Another closing php tag

**Review by adam-buckley — CHANGES_REQUESTED** [review]

File: system/html.php [hunk 3]
```diff
@@ -464,7 +464,9 @@ public static function form($data, $action = null, $method = "POST", $submitTitl
             $buffer .= "</div></div>";
         }
         $buffer .= "</div>";
-        $buffer .= "<script>$(function(){try{\$('textarea.ckeditor').each(function(){CKEDITOR.replace(this)})}catch(err){}});</script>";
+        // $buffer .= "<script>$(function(){try{\$('textarea.ckeditor').each(function(){CKEDITOR.replace(this)})}catch(err){}});</script>";
+
+        $buffer .= "<script>document.addEventListener('DOMContentLoaded', () => { [...document.querySelectorAll('textarea.ckeditor')].forEach(x => { CKEDITOR.replace(x) }) })</script>";
```
[the remark this answers is not in the document]
Comment (MaddyUnderStars) [reply system/html.php:469]: It returns an Array-like [NodeList](https://developer.mozilla.org/en-US/docs/Web/API/NodeList) structure which you cannot directly iterate over with .forEach

**Comment by adam-buckley** [comment]
image
Tests passing

File: system/templates/base/src/js/components/MultiSelect.ts
[the remark this answers is not in the document]
Comment (MaddyUnderStars) [reply system/templates/base/src/js/components/MultiSelect.ts:1]: Yes, the autocomplete class was just renamed from multiselect initially. It does all the same stuff except it also allows some things such as:
- autocomplete text input
- fetching data from some endpoint
- adding/removing data from the list via some endpoint
- thats all I can remember atm

**Review by iceaxeliehne — APPROVED** [review]

File: system/modules/auth/actions/login.php [hunk 4]
```diff
@@ -24,15 +26,15 @@ function login_POST(Web $w)
 
     $request_data = json_decode(file_get_contents("php://input"), true);
     if (empty($request_data) || !array_key_exists("login", $request_data) || !array_key_exists("password", $request_data)) {
-        $w->error("Please enter your login and password", "/auth/login");
+        $w->out((new JsonResponse())->setErrorResponse("Please enter your login and password", "Please enter your login and password"));
```
Comment (DerekCrannaford) [inline system/modules/auth/actions/login.php:29]: this $w->out::JSON strategy is known to make viewable JSON gibberish & error fragments in cases of low level exception events cascading from $w calls &/or hooks.
Can the entirety of POST have a try/catch wrapper with a stable-but-meaningless "System FAILED" JSON placeholder?

File: system/modules/auth/templates/login.tpl.php [hunk 5]
```diff
@@ -110,49 +112,49 @@
...
-                                document.getElementById("mfa_code").focus();
-                            });
-                        }
+                    method: 'POST',
+                    headers: new Headers({
+                        'Content-Type': 'application/json'
+                    }),
+                    body: JSON.stringify(formData)
+                }).then(response => {
+                    return response.json()
```
Comment (DerekCrannaford) [inline system/modules/auth/templates/login.tpl.php:123]: Maybe here is where to attend to : https://github.com/2pisoftware/cmfive-core/pull/389/files#r1868582045
???

File: system/modules/form/actions/application/attach_form.php [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    $application = FormApplicationService::getInstance($w)->getFormApplication($id);
+
+    if (empty($application->id)) {
+        $w->error("Application not found", '/form-application');
+        return;
+    }
+
+    $available_forms = FormService::getInstance($w)->getForms();
+
+    if (empty($available_forms)) {
```
Comment (DerekCrannaford) [inline system/modules/form/actions/application/attach_form.php:30]: This is an awkward bounce when in modal?

File: system/modules/form/actions/edit.php [hunk 7]
```diff
@@ -2,8 +2,10 @@
 
 function edit_GET(Web $w)
```
Comment (DerekCrannaford) [inline system/modules/form/actions/edit.php:3]: This wants to be in a modal, but it isn't?
image

image

File: system/modules/report/actions/createreport.php [hunk 8]
```diff
@@ -9,25 +14,60 @@ function createreport_ALL(Web &$w)
...
+            [new InputField([
+                "id|name" => "title",
+                "label" => "Title",
+                "value" => Request::string("title"),
+            ])],
+
+            [new Select([
+                "id|name" => "module",
+                "label" => "Module",
+                "value" => Request::string("module"),
```
Comment (MaddyUnderStars) [inline system/modules/report/actions/createreport.php:28]:
```suggestion
                "selected_option" => Request::string("module"),
```

File: system/modules/report/actions/createreport.php [hunk 9]
```diff
@@ -9,25 +14,60 @@ function createreport_ALL(Web &$w)
...
+                "label" => "Code",
+                "value" => Request::string("report_code"),
+                "rows" => 22,
+                "cols" => 110
+            ])],
+
+            [new Select([
+                "id|name" => "report_connection_id",
+                "label" => "Connect",
+                "value" => Request::int("report_connection_id"),
```
Comment (MaddyUnderStars) [inline system/modules/report/actions/createreport.php:56]:
```suggestion
                "selected_option" => Request::int("report_connection_id"),
```

File: system/modules/tag/partials/templates/listTags.tpl.php [hunk 10]
```diff
@@ -1,67 +1,71 @@
 <?php if (AuthService::getInstance($w)->user()->hasRole('tag_user') && $object->canView(AuthService::getInstance($w)->user())) : ?>
 
-<?php if ($object->canEdit(AuthService::getInstance($w)->user())) : ?>
-	<div id='tag_<?php echo get_class($object); ?>_<?php echo $object->id; ?>_modal' class='reveal-modal medium' data-reveal>
-		<a class="close-reveal-modal" aria-label="Close">&#215;</a>
-	</div>
-<?php endif; ?>
+    <style>
```
Comment (MaddyUnderStars) [inline system/modules/tag/partials/templates/listTags.tpl.php:3]: this embedded style should should be enqueued in the action instead

File: system/modules/task/actions/tasklist.php [hunk 11]
```diff
@@ -116,29 +118,62 @@ function tasklist_ALL(Web $w)
...
-        ])), // ["Task Group", "select", "task__task-group-id", !empty($task_group_id) ? $task_group_id : null, $taskgroup_data["taskgroups"]],
-        ["Task Type", "select", "task__type", !empty($task_type) ? $task_type : null, $taskgroup_data["types"]],
-        ["Task Priority", "select", "task__priority", !empty($filter_urgent) ? "Urgent" : (!empty($task_priority) ? $task_priority : null), $taskgroup_data["priorities"]],
-        ["Task Status", "select", "task__status", !empty($task_status) ? $task_status : null, $taskgroup_data["statuses"]],
-        (new \Html\Form\Select([
+
+        new Select([
+            "name|id" => "task__assignee-id",
+            "label" => "Assignee",
+            "value" => !empty($assignee_id) ? $assignee_id : null,
```
Comment (MaddyUnderStars) [inline system/modules/task/actions/tasklist.php:125]:
```suggestion
            "selected_option" => !empty($assignee_id) ? $assignee_id : null,
```

File: system/modules/task/actions/tasklist.php [hunk 12]
```diff
@@ -116,29 +118,62 @@ function tasklist_ALL(Web $w)
...
+            "options" => $filter_assignees,
+        ]),
+
+        new Select([
+            "name|id" => "task__creator-id",
+            "label" => "Creator",
+            "value" => !empty($creator_id) ? $creator_id : null,
```
Comment (MaddyUnderStars) [inline system/modules/task/actions/tasklist.php:132]:
```suggestion
            "selected_option" => !empty($creator_id) ? $creator_id : null,
```

File: system/modules/task/actions/tasklist.php [hunk 13]
```diff
@@ -116,29 +118,62 @@ function tasklist_ALL(Web $w)
...
+            "placeholder" => "Search",
+            "value" => !empty($task_group_id) ? $taskgroup->getSelectOptionValue() : null,
+            "source" => $w->localUrl("/task-group/ajaxAutocompleteTaskgroups"),
+            "minLength" => 2,
+        ])),
+
+        new Select([
+            "name|id" => "task__type",
+            "label" => "Task Type",
+            "value" => !empty($task_type) ? $task_type : null,
```
Comment (MaddyUnderStars) [inline system/modules/task/actions/tasklist.php:148]:
```suggestion
            "selected_option" => !empty($task_type) ? $task_type : null,
```

File: system/modules/task/actions/tasklist.php [hunk 14]
```diff
@@ -116,29 +118,62 @@ function tasklist_ALL(Web $w)
...
+            "options" => $taskgroup_data["types"],
+        ]),
+
+        new Select([
+            "name|id" => "task__priority",
+            "label" => "Task Priority",
+            "value" => !empty($filter_urgent) ? "Urgent" : (!empty($task_priority) ? $task_priority : null),
```
Comment (MaddyUnderStars) [inline system/modules/task/actions/tasklist.php:155]:
```suggestion
            "selected_option" => !empty($filter_urgent) ? "Urgent" : (!empty($task_priority) ? $task_priority : null),
```

File: system/modules/task/actions/tasklist.php [hunk 15]
```diff
@@ -116,29 +118,62 @@ function tasklist_ALL(Web $w)
...
+            "options" => $taskgroup_data["priorities"],
+        ]),
+
+        new Select([
+            "name|id" => "task__status",
+            "label" => "Task Status",
+            "value" => !empty($task_status) ? $task_status : null,
```
Comment (MaddyUnderStars) [inline system/modules/task/actions/tasklist.php:162]:
```suggestion
            "selected_option" => !empty($task_status) ? $task_status : null,
```

File: system/modules/task/actions/tasklist.php [hunk 16]
```diff
@@ -116,29 +118,62 @@ function tasklist_ALL(Web $w)
...
-            ["label" => "Yes", "value" => '1'],
-            ["label" => "Both", "value" => '']
-        ])->setSelectedOption($is_closed),
+            "id"        => "task__is_closed",
+            "options" => [
+                ["label" => "No", "value" => '0'],
+                ["label" => "Yes", "value" => '1'],
+                ["label" => "Both", "value" => '']
+            ],
+            "value" => $is_closed,
```
Comment (MaddyUnderStars) [inline system/modules/task/actions/tasklist.php:175]:
```suggestion
            "selected_option" => $is_closed,
```

File: system/templates/base/package.json [hunk 17]
```diff
@@ -1,51 +1,54 @@
...
-    "@codemirror/lang-html": "^6.4.5",
+    "@codemirror/lang-html": "^6.4.9",
     "@popperjs/core": "^2.11.8",
-    "bootstrap": "^5.2.3",
-    "bootstrap-icons": "^1.10.5",
+    "bootstrap": "^5.3.3",
+    "bootstrap-icons": "^1.11.3",
     "browser-process": "0.0.1",
     "codemirror": "^6.0.1",
     "jquery": "^3.7.1",
```
Comment (MaddyUnderStars) [inline system/templates/base/package.json:22]: should be removed?

**Review by MaddyUnderStars — CHANGES_REQUESTED** [review]
image
image
image

on form/show/:id event modal, select includes multiple "-- select --"

File: system/modules/auth/actions/login.php (hunk 4 above)
Comment (adam-buckley) [reply system/modules/auth/actions/login.php:29]: I'll put this in for post v6 changes

File: system/modules/auth/templates/profile.tpl.php [hunk 18]
```diff
@@ -1,153 +1,132 @@
...
-                <div class="small-12 medium-6 large-4 columns">
-                    <h3>Contact</h3>
-                    <label>Home Phone
-                        <?php
-                        echo (new \Html\Form\InputField\Tel([
+                            "class" => "form-control",
+                            "label" => "Language",
+                            "maxItems" => 1,
+                            "options" => [],
+                            "canCreate" => true, // this shouldn't be here
```
[the remark this answers is not in the document]
Comment (adam-buckley) [reply system/modules/auth/templates/profile.tpl.php:73]: Which one?

File: system/modules/file/templates/view.tpl.php [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+
+<?php
+if (!empty($owner)) {
+    $contact = $owner->getContact();
+    if (!empty($contact)) {
+        echo "Owner: " . ($contact->getFullName() ?: $contact->email);
+    } else {
+        // if there isn't a contact for this user
+        // fallback to the user id
+        echo "Owner: User (" . $user->id . ")";
```
[the remark this answers is not in the document]
Comment (adam-buckley) [reply system/modules/file/templates/view.tpl.php:12]: Probably not, something for post v6

File: system/modules/file/templates/view.tpl.php [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+        class="img-fluid"
+        data-caption="<?php echo $attachment->title; ?>"
+        src="<?php echo $attachment->getThumbnailUrl(); ?>" />
+<?php elseif ($attachment->isDocument()) :
+    echo $attachment->getDocumentEmbedHtml();
+else : ?>
+    <div
+        class="img-thumbnail
+            bi-filetype-doc
+            bi-filetype-<?php echo end(explode(".", $attachment->filename)) ?>
```
[the remark this answers is not in the document]
Comment (adam-buckley) [reply system/modules/file/templates/view.tpl.php:52]: Agreed, for post v6 as well

File: system/templates/base/package.json (hunk 17 above)
Comment (adam-buckley) [reply system/templates/base/package.json:22]: Browser-process? I'm not sure what purpose it serves, post v6 check

File: system/modules/auth/templates/profile.tpl.php (hunk 18 above)
[the remark this answers is not in the document]
Comment (MaddyUnderStars) [reply system/modules/auth/templates/profile.tpl.php:73]: the one in the highlighted code "this shouldn't be here". canCreate shouldn't be set to true for the profile language, it should have options set via I assume the lookup table?

File: system/templates/base/package.json (hunk 17 above)
Comment (MaddyUnderStars) [reply system/templates/base/package.json:22]: I meant the jquery dependency

**Comment by adam-buckley** [comment]
Oh right, not sure why that's there...

**Review by adam-buckley — APPROVED** [review]

File: system/templates/base/package.json (hunk 17 above)
Comment (adam-buckley) [reply system/templates/base/package.json:22]: Pushed to post v6

**Comment by MaddyUnderStars** [comment]
image
image
Codemirror dark mode colours aren't set. Also, opening a modal reattaches codemirror so you get 2 inputs.

**Review by MaddyUnderStars — APPROVED** [review]

**Pull request merged**